packages/grafana-ui/src/components/DataSourceSettings/CustomHeadersSettings.tsx

```diff
@@ -0,0 +1,198 @@
+      prevState => {
+        return { headers: [...prevState.headers, { name: 'X-Custom-Header', value: '', configured: false }] };
+      },
+      () => this.updateSettings(this.state.headers)
+    );
+  }
+
+  onHeaderChange(headerIndex: number, key: string, value: string) {
+    this.setState(prevState => {
+      return {
+        headers: prevState.headers.map((item, index) => {
+          if (headerIndex !== index) {
+            return item;
+          }
+          return {
+            ...item,
+            [key]: value,
+          };
+        }),
+      };
+    });
+  }
+
+  onHeaderReset(headerIndex: number) {
+    this.setState(prevState => {
+      return {
+        headers: prevState.headers.map((item, index) => {
+          if (headerIndex !== index) {
+            return item;
+          }
+          return {
+            ...item,
+            value: '',
+            configured: false,
+          };
+        }),
+      };
+    });
+  }
+
+  onHeaderRemove(headerIndex: number) {
+    this.setState(
+      prevState => {
+        return {
+          headers: [...prevState.headers.slice(0, headerIndex), ...prevState.headers.slice(headerIndex + 1)],
+        };
+      },
+      () => this.updateSettings(this.state.headers)
+    );
+  }
+
+  updateHeaders() {
+    const { jsonData, secureJsonData, secureJsonFields } = this.props.dataSourceConfig;
+    this.setState({
+      headers: Object.keys(jsonData)
+        .sort()
+        .filter(key => key.startsWith('httpHeaderName'))
+        .map((key, index) => {
+          return {
+            name: jsonData[key],
+            value: secureJsonData !== undefined ? secureJsonData[key] : '',
+            configured: (secureJsonFields && secureJsonFields[`httpHeaderValue${index + 1}`]) || false,
+          };
+        }),
+    });
+  }
+
+  render() {
+    const { headers } = this.state;
+
+    return (
+      <div className={'gf-form-group'}>
+        <div className="gf-form">
+          <h6>Custom HTTP Headers</h6>
+        </div>
+        <div>
+          {headers.map((header, index) => (
+            <div key={index} className="gf-form-inline">
+              <div className="gf-form">
+                <FormField
+                  label="Header"
+                  name="name"
+                  defaultValue={header.name}
+                  labelWidth={5}
+                  onChange={e => this.onHeaderChange(index, 'name', e.target.value)}
+                  onBlur={_ => this.updateSettings(this.state.headers)}
+                />
+              </div>
+              <div className="gf-form">
+                <SecretFormField
+                  label={'Value'}
+                  name={'value'}
+                  isConfigured={header.configured}
+                  value={header.value}
+                  labelWidth={5}
+                  inputWidth={header.configured ? 11 : 12}
+                  placeholder={'Header Value'}
+                  onReset={_ => this.onHeaderReset(index)}
```

Comment:
Instead of using `_` I suggest doing `() => ...`. This will remove the necessity for `onReset` type signature change in `SecretFormField`